scripts/recovery_sync

```diff
@@ -71,5 +71,6 @@ done
 rsync -a /etc/machine-id etc/
 rsync -a --relative /export/home/delphix/.ssh .
 
-find . -print0 | cpio --null --create --format=newc | gzip -7 >"${target}.tmp"
-mv "${target}.tmp" "$target"
+tmpfile=$(mktemp "${target}.XXXXX")
+find . -print0 | cpio --null --create --format=newc | gzip -7 >"$tmpfile"
+mv "$tmpfile" "$target"
```

Comment:
$target here is the one true `/boot/recovery.img` image, yes? If so, then is there still a race here? sync1 starts, gathers files into workdir1. Then, some other change happens and sync2 starts and gathers files into workdir2 (workdir1 is now stale). sync1 pauses for some reason, and sync2 finishes. `recovery.img` now contains the latest, correct, image (from workdir2). Now sync1 finally gets scheduled to finish and overwrites `recovery.img` with stale stuff. Is this possible? If so, should there be some higher-level locking involved?

Reply:
Using a file lock for the whole `recovery_sync` operation would likely be safer.

Reply:
I’ve used flock for this sort of thing in the upgrade scripts.. e.g. see [here](https://github.com/delphix/appliance-build/blob/8ce0ac130e725395efcf663a6e65b4b7be6494b3/upgrade/upgrade-scripts/execute#L44)